src/java/org/apache/cassandra/index/sai/disk/v2/V2VectorIndexSearcher.java

```diff
@@ -254,38 +253,39 @@ private SparseFixedBitSet bitSetForSearch()
     }
 
     @Override
-    public RangeIterator<PrimaryKey> limitToTopResults(QueryContext context, RangeIterator<Long> iterator, Expression exp, int limit) throws IOException
+    public RangeIterator limitToTopResults(QueryContext context, List<PrimaryKey> keys, Expression exp, int limit) throws IOException
     {
         try (PrimaryKeyMap primaryKeyMap = primaryKeyMapFactory.newPerSSTablePrimaryKeyMap())
         {
-            // the iterator represents keys from all the segments in our sstable -- we'll only pull of those that
+            // the iterator represents keys from the whole table -- we'll only pull of those that
             // are from our own token range so we can use row ids to order the results by vector similarity.
             var maxSegmentRowId = metadata.toSegmentRowId(metadata.maxSSTableRowId);
             SparseFixedBitSet bits = bitSetForSearch();
             var rowIds = new IntArrayList();
             try (var ordinalsView = graph.getOrdinalsView())
             {
-                while (iterator.hasNext())
+                for (PrimaryKey primaryKey : keys)
                 {
-                    Long sstableRowId = iterator.peek();
+                    if (primaryKey.compareTo(metadata.minKey) < 0 || primaryKey.compareTo(metadata.maxKey) > 0)
+                        continue;
+                    // TODO this is O(log n), is this a hotspot?
+                    long sstableRowId = primaryKeyMap.rowIdFromPrimaryKey(primaryKey);
```

Comment:
i think this is going to be a big deal since it's not just log n, it's disk-based log n.

IIRC this is why Zhao made the iterators generic -- so that you can leave the sstable postings as rowids, and you only have to deal with PK objects for the memtable data.

Thoughts:
(1) is there another way we can combine w/ memtable data w/o forcing everything to PK?
(2) alternatively, maybe we can make the iterators a Pair<PK, Long> so we can preserve rowids when we get them back from the sstable index instead of having to look them up again

Reply:
I agree it's likely a problem.

I have some relevant notes on this topic here and in the comments following: https://github.com/riptano/VECTOR-SEARCH/issues/14#issuecomment-1740377759

I already have a prototype of keeping track of `Pair<SSTableId, sstableRowId>` in PKs as we stream them, including merging PKs so we keep track of the relevant metadata. We could use that to limit the number of reads. However, the split row problem means that we need to search every matched PK against every vector index for the column on the table, so we'd still need to read from disk.

Reply:
One option is to get some performance tests to see how this change impacts us.

If we cannot find another solution, would it make sense to use a bloom filter?